BLSQ/iaso
# IA-2396: Payments part 2 - Lots

# Payment Lots API

This API allows for the management and querying of payment lots. Payment lots are collections of payments that can be processed together.

The Django model that stores "Payment Lot" is `PaymentLot`.

This API supports creating new payment lots, updating existing ones, and querying for payment lots based on various criteria such as creation date, status, and associated user.

## Permissions

- User must be authenticated
- User needs `iaso_payments` permission

## Status Computing

The status of a payment lot is dynamically computed based on the statuses of the payments it contains. The possible statuses are:

- `new`: Default status, indicating a newly created lot or a lot with no payments sent.
- `sent`: Indicates that all payments in the lot have been sent.
- `paid`: Indicates that all payments in the lot have been paid.
- `partially_paid`: Indicates that some, but not all, payments in the lot have been paid.

The status is computed every time a payment lot is saved, ensuring that the payment lot status accurately reflects the current state of its associated payments.

# Payment Lots Front-end:

- Allow user to select potential payments and create new lot of payments.
- Possibility to mark a lot as sent
- export list of payment to xlsx

Related JIRA tickets : IA-2396

## Self proofreading checklist


## Doc
- See swagger for api doc

## Changes

- use foreign keys to link change requests, payments and potential payments
- new api endpoint for lots 
- payments lots frontend (filters, ordering, mark as sent, export to xlsx)

## How to test

Make sure you have superuser right or iaso_payment permission.

You also need approved change request on org unit with the field created by well set.

Visit payments/to pay and create new lots of payments
Visit lots of payment page and play with filters, mark one as send and try to export it.

## Print screen / video


https://github.com/BLSQ/iaso/assets/12494624/e70ef5b9-068c-4a64-8856-5cb5fb1b4f02



## Notes

@mathvdh i could not add `mark_payments_as_sent` parameter on patch request in swagger doc, if you have an idea.
